scipy/optimize/_qap.py

```diff
@@ -0,0 +1,709 @@
+            used.
+            If `rng` is an int, a new ``RandomState`` instance is used,
+            seeded with `rng`.
+            If `rng` is already a ``RandomState`` or ``Generator``
+            instance, then that object is used.
+            Default is None.
+
+    Returns
+    -------
+    res : OptimizeResult
+        A :class:`scipy.optimize.OptimizeResult` containing the following
+        fields.
+
+        col_ind : 1-D array
+            An array of column indices corresponding with the best
+            permutation of the nodes of `B` found.
+        score : float
+            The corresponding value of the objective function.
+        nit : int
+            The number of Franke-Wolfe iterations performed during
+            the initialization resulting in the permutation
+            returned.
+
+    Notes
+    -----
+    The default method :ref:`'faq' <optimize.qap-faq>` uses the Fast
+    Approximate QAP algorithm [1]_; it is typically offers the best
+    combination of speed and accuracy.
+    Method :ref:`'2opt' <optimize.qap-2opt>` can be computationally expensive,
+    but may be a useful alternative, or it can be used to refine the solution
+    returned by another method.
+
+    References
+    ----------
+    .. [1] J.T. Vogelstein, J.M. Conroy, V. Lyzinski, L.J. Podrazik,
+           S.G. Kratzer, E.T. Harley, D.E. Fishkind, R.J. Vogelstein, and
+           C.E. Priebe, "Fast approximate quadratic programming for graph
+           matching," PLOS one, vol. 10, no. 4, p. e0121002, 2015,
+           https://doi.org/10.1371/journal.pone.0121002
+
+    .. [2] D. Fishkind, S. Adali, H. Patsolic, L. Meng, D. Singh, V. Lyzinski,
+           C. Priebe, "Seeded graph matching", Pattern Recognit. 87 (2019):
+           203-215, https://doi.org/10.1016/j.patcog.2018.09.014
+
+    .. [3] "2-opt," Wikipedia.
+           https://en.wikipedia.org/wiki/2-opt
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.optimize import quadratic_assignment
+    >>> cost = np.array([[0, 80, 150, 170], [80, 0, 130, 100],
+    ...         [150, 130, 0, 120], [170, 100, 120, 0]])
+    >>> dist = np.array([[0, 5, 2, 7], [0, 0, 3, 8],
+    ...         [0, 0, 0, 3], [0, 0, 0, 0]])
+    >>> res = quadratic_assignment(cost, dist)
+    >>> print(res)
+     col_ind: array([0, 3, 2, 1])
+         nit: 9
+       score: 3260
+
+    The see the relationship between the returned ``col_ind`` and ``score``,
+    use ``col_ind`` to form the best permutation matrix found, then evaluate
+    the objective function :math:`f(P) = trace(A^T P B P^T )`.
+
+    >>> n = cost.shape[0]
+    >>> perm = res['col_ind']
+    >>> P = np.eye(n)[perm]
+    >>> score = int(np.trace(cost.T @ P @ dist @ P.T))
+    >>> print(score)
+    3260
+
+    Alternatively, to avoid constructing the permutation matrix explicitly,
+    directly permute the rows and columns of the distance matrix.
+
+    >>> score = np.trace(cost.T @ dist[np.ix_(perm, perm)])
```

Comment:
I find `A[P][:, P]` more intuitive than  `A[np.ix_(P, P)]`. The latter requires a user to look up a fairly obscure numpy function. I didn't know it.

On my laptop the former is also faster for `len(A) <= 800`.

Reply:
OK changed.